### PR TITLE
Add the temperature sensor, only available on Number1 for now.

### PR DESCRIPTION
**Pull request by nomis52**

### CODE REVIEW

**Comment by nomis52** [comment]
Closes #206

File: firmware/src/app.c [hunk 1]
```diff
@@ -158,6 +161,7 @@ void APP_Tasks(void) {
   if (Transceiver_GetMode() == T_MODE_RESPONDER) {
     RDMHandler_Tasks();
     SPIRGB_Tasks();
+    Temperature_Tasks();
```
Comment (peternewman) [inline firmware/src/app.c]: Is it worth it being available in all modes for status reporting?

File: boardcfg/ethernet_sk2/app_settings.h [hunk 2]
```diff
@@ -141,6 +141,13 @@ extern "C" {
 #define RDM_RESPONDER_MUTE_PORT_BIT PORTS_BIT_POS_12
 
 /**
+ * @brief The pin to use for the RDM temperature sensor.
+ *
+ * If undefined, we use a 'dummy' sensor that generates random values.
```
Comment (peternewman) [inline boardcfg/ethernet_sk2/app_settings.h]: Aren't we better off not doing an alternative dummy sensor and having less sensors, so we can differentiate if it's working or not?

File: firmware/src/sensor_model.c [hunk 3]
```diff
@@ -182,15 +194,15 @@ static const ProductDetailIds PRODUCT_DETAIL_ID_LIST = {
 static const SensorDefinition SENSOR_DEFINITIONS[] = {
   {
     .description = SENSOR_NAME1,
-    .normal_maximum_value = 50,
-    .normal_minimum_value = 0,
-    .range_maximum_value = 100,
-    .range_minimum_value = -10,
+    .normal_maximum_value = 500,
+    .normal_minimum_value = 300,
```
Comment (peternewman) [inline firmware/src/sensor_model.c]: It reads about 40 Celsius normally then?

File: firmware/src/temperature.c [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+static CoarseTimer_Value g_timer;
+
+struct {
+  bool new_sample;  //!< true if there is a new sample
+  uint16_t sample_value;  //!< The raw sampled value
+  uint16_t temperature;  //!< The temperature in 10ths of a degree.
+} g_adc_data;
+
+void __ISR(_ADC_VECTOR, ipl1AUTO) ADCEvent() {
+  // AD1CON1bits.ASAM = 0;
```
Comment (peternewman) [inline firmware/src/temperature.c]: Remove

File: firmware/src/temperature.c [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+  PLIB_ADC_SampleAutoStartDisable(ADC_ID_1);
+  SYS_INT_SourceStatusClear(INT_SOURCE_ADC_1);
+  SYS_INT_SourceDisable(INT_SOURCE_ADC_1);
+  g_adc_data.new_sample = true;
+
+  // g_adc_data.value = ADC1BUF0;
```
Comment (peternewman) [inline firmware/src/temperature.c]: Remove

File: firmware/src/temperature.c [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+#ifdef RDM_RESPONDER_TEMPERATURE_SENSOR
+  g_timer = CoarseTimer_GetTime();
+  g_adc_data.new_sample = false;
+  g_adc_data.sample_value = 0;
+  g_adc_data.temperature = 0;
+
+  // The pin must be in analog mode - configure this with the harmony
+  // configurator plugin.
+
+  // AD1CON1 = 0x00e0;  // auto conversion, 16bit int
```
Comment (peternewman) [inline firmware/src/temperature.c]: Remove

File: firmware/src/temperature.c [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+  PLIB_ADC_ResultFormatSelect(ADC_ID_1, ADC_RESULT_FORMAT_INTEGER_16BIT);
+  PLIB_ADC_ConversionTriggerSourceSelect(
+      ADC_ID_1, ADC_CONVERSION_TRIGGER_INTERNAL_COUNT);
+
+  // AD1CHSbits.CH0SA = 2;
```
Comment (peternewman) [inline firmware/src/temperature.c]: Remove

File: firmware/src/temperature.c [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+  PLIB_ADC_MuxChannel0InputPositiveSelect(
+      ADC_ID_1, ADC_MUX_A, ADC_INPUT_POSITIVE_AN2);
+
+  // AD1CSSL = 0;  // no scanning mask
```
Comment (peternewman) [inline firmware/src/temperature.c]: Remove

File: firmware/src/temperature.c [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+
+  // The only method Harmony 1.06 provides to set ADCS uses the bus speed, so
+  // rather than depend on that, we set it here manually.
+  AD1CON3 = 0x0002;  // ADCS
+  PLIB_ADC_ConversionClockSourceSelect(
+      ADC_ID_1, ADC_CLOCK_SOURCE_SYSTEM_CLOCK);
+  PLIB_ADC_SampleAcquisitionTimeSet(ADC_ID_1, 0x1f);  // 31 x Tad
```
Comment (peternewman) [inline firmware/src/temperature.c]: What's Tad?

File: firmware/src/temperature.c [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+
+  // AD1CON2 = 0;  // VDD & VSS, no scan, mux A
```
Comment (peternewman) [inline firmware/src/temperature.c]: Remove

File: firmware/src/temperature.c [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+  PLIB_ADC_MuxAInputScanDisable(ADC_ID_1);
+  PLIB_ADC_SamplesPerInterruptSelect(ADC_ID_1, ADC_1SAMPLE_PER_INTERRUPT);
+
+  // AD1CON1bits.ADON = 1;
```
Comment (peternewman) [inline firmware/src/temperature.c]: Remove

**Comment by peternewman** [comment]
Generally LGTM, just a few comments.

File: boardcfg/ethernet_sk2/app_settings.h (hunk 2 above)
Comment (nomis52) [inline boardcfg/ethernet_sk2/app_settings.h]: The dummy sensor tests other things as well, like prefix handling.

File: firmware/src/app.c (hunk 1 above)
Comment (nomis52) [inline firmware/src/app.c]: We can add this later when / if we need it.

File: firmware/src/sensor_model.c (hunk 3 above)
Comment (nomis52) [inline firmware/src/sensor_model.c]: 40 to 45

File: firmware/src/temperature.c (hunk 4 above)
Comment (nomis52) [inline firmware/src/temperature.c]: The ADC has a number of different modes and requires multiple registers to be configured. I've written out the config values here so when we next have to change the code we know what's going on. 

Between 1.02 and 1.06 some of the harmony function names changed. Being able to map to registers is useful.

File: firmware/src/temperature.c (hunk 9 above)
Comment (nomis52) [inline firmware/src/temperature.c]: The period of the A2D unit. It's a multiple of the peripheral bus period.

File: boardcfg/ethernet_sk2/app_settings.h (hunk 2 above)
Comment (peternewman) [inline boardcfg/ethernet_sk2/app_settings.h]: But I mean I don't think it's possible to tell if the sensor is real or not if it always acts the same, perhaps switch it to amps or something if it's not a genuine reading?

File: firmware/src/sensor_model.c (hunk 3 above)
Comment (peternewman) [inline firmware/src/sensor_model.c]: Okay

File: firmware/src/temperature.c (hunk 4 above)
Comment (peternewman) [inline firmware/src/temperature.c]: Fair enough

File: firmware/src/temperature.c (hunk 9 above)
Comment (peternewman) [inline firmware/src/temperature.c]: Ah okay

File: firmware/src/temperature.c (hunk 4 above)
Comment (peternewman) [inline firmware/src/temperature.c]: Perhaps just add a comment at the top explaining this?

**Pull request merged**